subsys/testsuite/ztest/src/ztest_new.c

```diff
@@ -555,12 +543,21 @@ static int z_ztest_run_test_suite_ptr(struct ztest_suite_node *suite)
 		fail = 0;
 
 #ifdef CONFIG_ZTEST_SHUFFLE
-		z_ztest_shuffle(_ztest_unit_test_list_start,
-				_ztest_unit_test_list_end - _ztest_unit_test_list_start, &tmp,
-				sizeof(struct ztest_unit_test));
-#endif
 
+		struct ztest_unit_test *tests_to_run[ZTEST_TEST_COUNT];
+
+		memset(tests_to_run, 0, ZTEST_TEST_COUNT * sizeof(struct ztest_unit_test *));
+		z_ztest_shuffle((void **)tests_to_run, (intptr_t)_ztest_unit_test_list_start,
+				ZTEST_TEST_COUNT, sizeof(struct ztest_unit_test));
+		for (size_t i = 0; i < ZTEST_TEST_COUNT; ++i) {
+			test = tests_to_run[i];
+			/* Make sure that the test belongs to this suite */
+			if (strcmp(suite->name, test->test_suite_name) != 0) {
+				continue;
+			}
+#else
 		while (((test = ztest_get_next_test(suite->name, test)) != NULL)) {
```

Comment:
We can update ztest_get_next_test to accept the head pointer of the list, so we don't duplicate code here.

Reply:
The problem is that we have two different lists, `struct ztest_unit_test *[]` and `struct ztest_unit_test *` so I'm not sure there's a good way for it. I split them up so the whole loop body is in a single `#if`, should make it easier to understand